### ISSUE TITLE
Building Angular: Yarn link to project

### ISSUE THREAD

**Issue by s-gbz** [angular/angular#31908]
# 📚 Docs or angular.io bug report

### Description
When working on a codebase most developers appreciate being able to test and experience the changes they commit. In order to ease new Angular contributors it would be most helpful to have some documentation how to use a local Angular build as a dependecy in a test project.

I think it would be great if someone more experienced could add some exemplaric steps to the official contriubuting guide.

### What's the affected URL?**
[DEVELOPER.md#building / #running tests locally](https://github.com/angular/angular/blob/master/docs/DEVELOPER.md#building)

### Reproduction Steps**
So far I've
- build Angular via `angular/scripts/build-packages.sh`
- run `yarn link` inside `angular/dist/package-dist/router` to create a symlink for `@angular/router`
- and tried using it in my test project via `yarn link @angular/router` 

After many different approaches (for example linking the contents of `angular/dist/bin/packages/router`) it never worked, because the routing module itself or other dependencies like `@angular/common/core/platform-browser`/... could not be found.

### Expected vs Actual Behavior**
Expected to have more detailed instructions how to link the contents of the `dist/packages-dist ` folder to a test project (similiar to the subprojects in `angular/integration`).

## 📷Screenshot
![angular-link-error](https://user-images.githubusercontent.com/23424538/62128238-47dd3100-b2d4-11e9-84b5-e7c664b885a1.PNG)

## 🔥 Exception or Error
<pre><code>
ERROR in ../angular/dist/router/fesm5/router.js
Module not found: Error: Can't resolve '@angular/common' in 'C:\Users\grilborzer\Code\angular\dist\router\fesm5'
ERROR in ../angular/dist/router/fesm5/router.js
Module not found: Error: Can't resolve '@angular/core' in 'C:\Users\grilborzer\Code\angular\dist\router\fesm5'
ERROR in ../angular/dist/router/fesm5/router.js
Module not found: Error: Can't resolve '@angular/platform-browser' in 'C:\Users\grilborzer\Code\angular\dist\router\fesm5'
</code></pre>

## 🌍  Your Environment
Windows 10 [Version 10.0.16299.1268] Enterprise v. 1709
Yarn 1.16.0
Node 10.9.0

### Anything else relevant?
I tried mentioned steps on Windows and on an Ubuntu Subsystem.

**Comment by s-gbz** [comment]
@kapunahelewong may I ask if there's any chance on this issue being addressed some time soon?
I see that the backlog is quite busy. Are there any other resource I could advise in the mean time? The point is that the issue is holding me back from working on other issues in the angular repository :(

**Comment by kapunahelewong** [comment]
You're right, there are a lot of items in the backlog. I'm not sure what the solution is for the issue. Maybe @petebacondarwin or @gkalpak could advise us?

**Comment by gkalpak** [comment]
I haven't tried `yarn link` with the built Angular packages. (I suspect it might not work anyway.)
What works for me for testing the built packages against a local project is changing the project's `package.json` to point to the local Angular packages.

E.g. if I have built the packages in `/angular/dist/packages-dist/` and I want to test them on a project at `/my-projects/test-project/`, I would update `/my-projects/test-project/package.json` as follows:

```diff
 {
   ...
   "dependencies": {
-    "@angular/common": "9.0.0",
+    "@angular/common": "file:../../angular/dist/packages-dist/common",
-    "@angular/core": "9.0.0",
+    "@angular/core": "file:../../angular/dist/packages-dist/core",
     ...
   }
 }
```

After that, running `npm install` or `yarn install` will install the locally built Angular packages.

LMK if this is enough to unblock you, @s-gbz.

##
BTW, since recently, it is also possible to grab the packages from a PR's [build artifacts][1]. 

[1]: https://github.com/angular/angular/blob/31dccab2/docs/DEVELOPER.md#getting-packages-from-build-artifacts

**Comment by s-gbz** [comment]
Looks like it worked!! (I think I had tried sth. similiar, too bad it didn't work back then) Thank you so much @gkalpak ❤️

**Comment by llmartinll** [comment]
https://angular.io/guide/creating-libraries seems to suggest it/s possible, did anyone manage to get it working?

Spent quite some time today attempting to use the built package of my Angular Libary's `dist` project in a separate Angular project that imports the library via `yarn link` but could not get it working.

 @gkalpak mentioned solution works, but it would be very nice to be able to work in the libary and see the changes real-time by using `ng build --prod --watch` for the library project and the `yarn link`'s symlink to import the built files into your app.

**Comment by s-gbz** [comment]
This feel like an eternity ago, but yeah the solution George provided had worked for me 👍 
I'm afraid that including real-time changes is rather hard to work out, because the build process takes some time itself. 

I'd suggest you try using artifacts and most importantly include the proper **peer dependencies**.
Here's a `package.json` that works for me:

```json
{
  "name": "angular-test",
  "version": "0.0.0",
  "scripts": {
    "ng": "ng",
    "start": "ng serve",
    "build": "ng build",
    "test": "ng test",
    "lint": "ng lint",
    "e2e": "ng e2e"
  },
  "private": true,
  "dependencies": {
    "@angular/animations": "~7.2.0",
    "@angular/common": "https://xxx-yyy-gh.circle-artifacts.com/0/angular/common-pr33572-40677d5.tgz",
    "@angular/compiler": "~7.2.0",
    "@angular/core": "https://xxx-yyy-gh.circle-artifacts.com/0/angular/core-pr33572-40677d5.tgz",
    "@angular/forms": "~7.2.0",
    "@angular/platform-browser": "https://xxx-yyy-gh.circle-artifacts.com/0/angular/platform-browser-pr33572-40677d5.tgz",
    "@angular/platform-browser-dynamic": "https://xxx-yyy-gh.circle-artifacts.com/0/angular/platform-browser-dynamic-pr33572-40677d5.tgz",
    "@angular/router": "~7.2.0",
    "core-js": "^2.5.4",
    "rxjs": "~6.3.3",
    "tslib": "^1.9.0",
    "zone.js": "~0.8.26"
  },
  "peerDependencies": {
    "@angular/core": "7.2.15",
    "@angular/common": "7.2.15",
    "@angular/platform-browser": "7.2.15",
    "@angular/compiler": "https://xxx-yyy-gh.circle-artifacts.com/0/angular/compiler-pr33572-40677d5.tgz",
    "zone.js": "0.10.2",
    "rxjs": "~6.3.3"
  },
  "devDependencies": {
    "@angular-devkit/build-angular": "~0.13.0",
    "@angular/cli": "~7.3.8",
    "@angular/compiler-cli": "~7.2.0",
    "@angular/language-service": "~7.2.0",
    "@types/node": "~8.9.4",
    "@types/jasmine": "~2.8.8",
    "@types/jasminewd2": "~2.0.3",
    "codelyzer": "~4.5.0",
    "jasmine-core": "~2.99.1",
    "jasmine-spec-reporter": "~4.2.1",
    "karma": "~4.0.0",
    "karma-chrome-launcher": "~2.2.0",
    "karma-coverage-istanbul-reporter": "~2.0.1",
    "karma-jasmine": "~1.1.2",
    "karma-jasmine-html-reporter": "^0.2.2",
    "protractor": "~5.4.0",
    "ts-node": "~7.0.0",
    "tslint": "~5.11.0",
    "typescript": "~3.2.2"
  }
}
```

**Comment by cerkiner** [comment]
> I haven't tried `yarn link` with the built Angular packages. (I suspect it might not work anyway.)
> What works for me for testing the built packages against a local project is changing the project's `package.json` to point to the local Angular packages.
> 
> E.g. if I have built the packages in `/angular/dist/packages-dist/` and I want to test them on a project at `/my-projects/test-project/`, I would update `/my-projects/test-project/package.json` as follows:
> 
> ```diff
>  {
>    ...
>    "dependencies": {
> -    "@angular/common": "9.0.0",
> +    "@angular/common": "file:../../angular/dist/packages-dist/common",
> -    "@angular/core": "9.0.0",
> +    "@angular/core": "file:../../angular/dist/packages-dist/core",
>      ...
>    }
>  }
> 
> 
> After that, running `npm install` or `yarn install` will install the locally built Angular packages.
> 
> LMK if this is enough to unblock you, @s-gbz.
> 
> BTW, since recently, it is also possible to grab the packages from a PR's [build artifacts](https://github.com/angular/angular/blob/31dccab2/docs/DEVELOPER.md#getting-packages-from-build-artifacts).

Installation is successful but I am facing another issue when serving the test project;
`An unhandled exception occurred: Invalid Version: 9.1.9.with-local-changes`

**Comment by s-gbz** [comment]
I'm sorry, I haven't been up to date with Angular contributions and thus can't help you out here.
I'll try to comeback if I have any updates. 

You might try https://github.com/whitecolor/yalc in the meantime.
Haven't tried the tool myself, but it seems to be pretty much what we need 🚀 🙏
